facilitator/src/transport/gcs.rs

```diff
@@ -373,98 +383,81 @@ impl StreamingTransferWriter {
-                    return Err(anyhow!(
-                        "No range header in response from GCS: {:?}",
-                        http_response.into_string()
-                    ));
-                }
                 let range_header = http_response.header("Range").unwrap();
-                // The range header is like "bytes=111-222", and represents a
-                // range in the overall object, not the current chunk.
-                let range = range_header
-                    .strip_prefix("bytes=")
-                    .context(format!("unexpected Range header {}", range_header))?;
-                let mut values = range.splitn(2, '-');
-                let start = values
-                    .next()
-                    .context(format!("unexpected Range header {}", range_header))?
+                // The range header is like "bytes=0-222", and represents the
+                // uploaded portion of the overall object, not the current chunk
+                let end = range_header
+                    .strip_prefix("bytes=0-")
+                    .context(format!(
+                        "Range header {} missing bytes prefix",
+                        range_header
+                    ))?
                     .parse::<usize>()
-                    .context(format!("unexpected Range header {}", range_header))?;
-                let end = values
-                    .next()
-                    .context(format!("unexpected Range header {}", range_header))?
-                    .parse::<usize>()
-                    .context(format!("unexpected Range header {}", range_header))?;
-
-                if start != self.object_upload_position {
-                    return Err(anyhow!("unexpected Range header {}", range_header));
-                }
-
-                self.object_upload_position = end + 1;
+                    .context("End in range header {} not a valid usize")?;
 
                 // If we have a little content left over, we can't just make
                 // another request, because if there's too little of it, Google
-                // will reject it. Instead, put the portion of the chunk that we
-                // didn't manage to upload back into self.buffer so it can be
+                // will reject it. Instead, leave the portion of the chunk that
+                // we didn't manage to upload back in self.buffer so it can be
                 // handled by a subsequent call to upload_chunk.
-                self.buffer
-                    .extend_from_slice(&chunk[self.object_upload_position - start..]);
-            }
-            _ => {
-                return Err(anyhow!(
-                    "failed to upload part to GCS: {} synthetic: {}\n{:?}",
-                    http_response.status(),
-                    http_response.synthetic(),
-                    http_response.into_string()
-                ))
+                let new_buf = self.buffer.split_off(end + 1 - self.object_upload_position);
```

Comment:
Here, and below when we update object_upload_position, we are giving a lot of trust to the `end` value read from the network. A malicious value could crash us. Before we trust it, we should check that it's reasonable - greater than zero and less than or equal to `object_length_thus_far`.